org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/LambdaExpression.java

```diff
@@ -560,9 +560,23 @@ private void analyzeExceptions() {
 		boolean oldAnalyseResources = compilerOptions.analyseResourceLeaks;
 		compilerOptions.analyseResourceLeaks = false;
 		try {
+			int firstLocalLocal = this.scope.outerMostMethodScope().analysisIndex;
+			LocalVariableBinding[] locals = this.scope.locals;
+			if (locals != null) {
+				for (LocalVariableBinding local : locals) {
+					if (local != null) {
+						if (local.isParameter()) {
```

Comment:
Why is this distinction necessary? Is it because parameters are never unassigned? But treating them the same as regular locals wouldn't hurt, would it?

Reply:
For "regular" non-parameter locals of the current lambda - we don't want to mark them as definitely assigned. That was the bug here - we are assigning a blank final local as being already definitely assigned and when we see an assignment in the body of the lambda, report an error about double assignment to final variable - this inhibits exception analysis since we flag the lambda as having ignored mandatory errors.

So this code is computing the high water mark index value up to which we want to mark the locals (outer locals + parameters of the current lambda if any) as definitely assigned.

Reply:
Sure, I just figured that parameters are always marked as definitely assigned anyway, and hence they don't _need_ the additional marking (which would simplify the code). In other words: only variables from truly outer scopes need the special treatment. Or am I missing smth?

Reply:
> Sure, I just figured that parameters are always marked as definitely assigned anyway, and hence they don't _need_ the additional marking (which would simplify the code). In other words: only variables from truly outer scopes need the special treatment. Or am I missing smth?

**Here**, we **start** the analysis at the lambda **body** - 

c.f. 

`org.eclipse.jdt.internal.compiler.ast.LambdaExpression.analyseCode(BlockScope, FlowContext, FlowInfo)` which by calling `org.eclipse.jdt.internal.compiler.ast.AbstractMethodDeclaration.analyseArguments(LookupEnvironment, FlowInfo, Argument[], MethodBinding)` ensures that parameters are marked as definitely assigned. 

That part of the analysis dealing with the lambda header is skipped here, so this is not *additional* marking, but markiking *in lieu*

Reply:
> **Here**, we **start** the analysis at the lambda **body** -

I see, so we bypass `analyseArguments`. Thanks for explaining.